### PR TITLE
[PHP 8.1] Add support for values method in SpatieEnumClassToEnumRector

### PR DESCRIPTION
**Pull request by JoolsMcFly**
Hi guys.

I added a fixture that includes value mapping as it is not supported by the current implementation.

I've only started using Rector recently and this is my first PR. It gets the job done (but fails on some phpstan rules) but I am certain code can be better. Any pointer to improve my code is more than welcome!

Thanks for checking it out!

### CODE REVIEW

**Comment by samsonasik** [comment]
Github action seems not immediatelly started, let's close and re-open

File: rules/Php81/NodeFactory/EnumFactory.php [hunk 1]
```diff
@@ -91,14 +96,76 @@ private function createEnumCaseFromConst(ClassConst $classConst): EnumCase
...
+
+    private function getValuesClassMethod(Class_ $class): ?ClassMethod
+    {
+        foreach ($class->stmts as $stmt) {
+            if ($stmt instanceof ClassMethod && $stmt->name->name === 'values') {
+                return $stmt;
+            }
+        }
+
+        return null;
```
Comment (samsonasik) [inline rules/Php81/NodeFactory/EnumFactory.php]: You can directly `return $class->getMethod('values')`

```suggestion
        return $class->getMethod('values');
```

**Comment by samsonasik** [comment]
PHPStan notice need to be resolved to make CI green https://github.com/rectorphp/rector-src/actions/runs/3696257400/jobs/6259717254#step:5:10

File: rules/Php81/NodeFactory/EnumFactory.php [hunk 2]
```diff
@@ -91,14 +96,76 @@ private function createEnumCaseFromConst(ClassConst $classConst): EnumCase
...
     }
+
+    /**
+     * @return array<string, int|string>
+     */
+    private function generateMappingFromClass(Class_ $class): array
+    {
+        $mapping = [];
+
+        $classMethod = $this->getValuesClassMethod($class);
```
Comment (samsonasik) [inline rules/Php81/NodeFactory/EnumFactory.php]: You can directly use:

```suggestion
        $classMethod = $class->getMethod('values');
```

File: rules/Php81/NodeFactory/EnumFactory.php (hunk 1 above)
Comment (JoolsMcFly) [reply rules/Php81/NodeFactory/EnumFactory.php]: Excellent, thanks!

File: rules/Php81/NodeFactory/EnumFactory.php [hunk 3]
```diff
@@ -91,14 +98,70 @@ private function createEnumCaseFromConst(ClassConst $classConst): EnumCase
...
+     * @return array<int|string, mixed>
+     */
+    private function generateMappingFromClass(Class_ $class): array
+    {
+        $mapping = [];
+
+        $classMethod = $class->getMethod('values');
+        if ($classMethod === null || !is_array($classMethod->stmts)) {
+            return $mapping;
+        }
```
Comment (samsonasik) [inline rules/Php81/NodeFactory/EnumFactory.php]:
```suggestion
        if (! $classMethod instanceof ClassMethod) {
           return [];
        }
```

File: rules/Php81/NodeFactory/EnumFactory.php [hunk 4]
```diff
@@ -91,14 +98,70 @@ private function createEnumCaseFromConst(ClassConst $classConst): EnumCase
...
+
+        foreach ($classMethod->stmts as $methodStmt) {
```
Comment (samsonasik) [inline rules/Php81/NodeFactory/EnumFactory.php]:
```suggestion
        foreach ((array) $classMethod->stmts as $methodStmt) {
```

File: rules/Php81/NodeFactory/EnumFactory.php [hunk 5]
```diff
@@ -91,14 +98,70 @@ private function createEnumCaseFromConst(ClassConst $classConst): EnumCase
...
+            if (
+                !($methodStmt instanceof Return_)
+                || !($methodStmt->expr instanceof Expr\Array_)
+            ) {
+                continue;
+            }
```
Comment (samsonasik) [inline rules/Php81/NodeFactory/EnumFactory.php]: What if there are `Return_` inside of `If_`? I think you can use `BetterNodeFinder->findInstancesOfInFunctionLikeScoped()`

```php
$returns = $this->betterNodeFinder->findInstancesOfInFunctionLikeScoped($classMethod, Return_::class);
```

and then loop the returns instead.

Comment (samsonasik) [reply rules/Php81/NodeFactory/EnumFactory.php]: I see, the `if` condition possibly never happen on enum class `values` method :)

File: rules/Php81/NodeFactory/EnumFactory.php [hunk 6]
```diff
@@ -91,14 +98,70 @@ private function createEnumCaseFromConst(ClassConst $classConst): EnumCase
...
 
         return new EnumCase($enumName, $enumExpr);
     }
+
+    /**
+     * @return array<int|string, mixed>
+     */
+    private function generateMappingFromClass(Class_ $class): array
+    {
+        $mapping = [];
```
Comment (samsonasik) [inline rules/Php81/NodeFactory/EnumFactory.php]: this can be moved after next if.

Comment (JoolsMcFly) [reply rules/Php81/NodeFactory/EnumFactory.php]: Ive just committed some changes to address your comments. Thanks.

File: rules/Php81/NodeFactory/EnumFactory.php (hunk 5 above)
Comment (JoolsMcFly) [reply rules/Php81/NodeFactory/EnumFactory.php]: Nice one.

**Comment by samsonasik** [comment]
Please update PR title, to something like:

```
[PHP 8.1] Add values method return to case enum support on SpatieEnumClassToEnumRector
```

**Comment by JoolsMcFly** [comment]
@samsonasik  Just changed the PR title to something more meaningful, slightly different to what you suggested. Is it ok?

**Comment by samsonasik** [comment]
Double quote in "values" can be removed from PR title, it was cause error in build scoped in the past when PR merged with squash include PR title.

**Comment by JoolsMcFly** [comment]
I see. PR tile is now quotes free!

**Review by samsonasik — APPROVED** [review]
Looks good

**Comment by TomasVotruba** [comment]
Thank you :clap:

**Comment by JoolsMcFly** [comment]
While on the subject I was planning on replacing usages of Spatie enums with native enums because current Rector rule only transforms the enum. In our project I had to replace all usages manually and it is both time consuming and error prone.

For example we need to change
* `$statusEnum::draft()->getValue()` to `$statusEnum::DRAFT->value`.
* `$status->equals(StatusEnum::draft());` to `$status === StatusEnum::draft();`
* etc

Would that be a good candidate for my next PR? I can split replacing of usages into several PRs to keep them small and easier to review. WDYT?

**Pull request merged**